In [46]:
import numpy as np
import pandas as pd
import seaborn as sns
from datetime import date

# Tasca M3 T02

## Ejercicio 1
Descargue el dataset *Airlines Delay: Airline donde-time statistics and delay causas* y cárguelo a un Pandas Dataframe. Explore los datos que contiene, explique brevemente qué variables hay y quédese únicamente con las columnas que considere relevantes. Justifique su elección.

### Importante

Reduzca la dimensión del dataset de manera aleatoria para obtener un dataset de solo 200.000 registros. Todos los ejercicios se tienen que hacer con este dataset reducido.

In [5]:
df=pd.read_csv('DelayedFlights.csv')

In [47]:
df_flights=df.sample(n=200000)
df_flights.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
365983      1148586  2008      2          26          2   1321.0        1300   
167859       544978  2008      1          28          1   1909.0        1858   
1203373     4020244  2008      7          31          4   1513.0        1445   
1777761     6566770  2008     12          18          4   1832.0        1822   
432034      1336044  2008      3          15          6   2005.0        1520   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
365983    1739.0        1718            CO  ...    13.0    18.0          0   
167859    2248.0        2238            AS  ...     8.0    20.0          0   
1203373   1639.0        1530            MQ  ...     5.0    35.0          0   
1777761   2036.0        2020            XE  ...    21.0    11.0          0   
432034    2143.0        1646            OH  ...    10.0    16.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
365983                  N         0          21.0           0.0      0.0   
167859                  N         0           NaN           NaN      NaN   
1203373                 N         0           0.0           0.0     49.0   
1777761                 N         0          10.0           0.0      6.0   
432034                  N         0         285.0           0.0     12.0   

        SecurityDelay  LateAircraftDelay  
365983            0.0                0.0  
167859            NaN                NaN  
1203373           0.0               20.0  
1777761           0.0                0.0  
432034            0.0                0.0  

[5 rows x 30 columns]

* Primero haremos una tabla para organizar nuestras variables.

$$
\begin{aligned}
& \text {Información de Fechas }\\
&\begin{array}{|l|r|}
 \hline\text{Year } & \text { Año}\\
 \hline\text{Month } & \text {  Mes }\\
 \hline\text{DayofMonth } & \text { Día del mes }\\
 \hline\text{DayofWeek } & \text { Día de la semana }\\
\hline
\end{array}
\end{aligned}
$$

$$
\begin{aligned}
& \text {Información de Salida / minutos }\\
&\begin{array}{|l|r|}
 \hline\text{DepTime } & \text { Hora de salia }\\
 \hline\text{CRSDepTime } & \text { Hora de salida programada }\\
 \hline\text{TaxiOut} & \text { Tiempo desde que el avión está rodando hasta que sale de pista}\\
  \hline\text{DepDelay } & \text { Retraso de salida }\\
 \hline
\end{array}
\end{aligned}
$$

$$
\begin{aligned}
& \text {Información de Llegada / minutos}\\
&\begin{array}{|l|r|}
 \hline\text{ArrTime } & \text { Hora de llegada }\\
 \hline\text{CRSArrTime } & \text { Hora de llegada programada }\\
 \hline\text{TaxiIn } & \text {Tiempo desde que el avión aterriza hasta que llega a Gate }\\
 \hline\text{ArrDelay } & \text { Retraso de llegada }\\
 \hline
\end{array}
\end{aligned}
$$

$$
\begin{aligned}
& \text {Información Aerolínea}\\
&\begin{array}{|l|r|}
 \hline\text{UniqueCarrier } & \text { Código único de la compañía aérea }\\
 \hline\text{FlightNum } & \text { Número de vuelo }\\
 \hline\text{TailNum } & \text { Número de cola de la aeronave }\\
 \hline
\end{array}
\end{aligned}
$$

$$
\begin{aligned}
& \text {Causa de los retrasos}\\
&\begin{array}{|l|r|}
\hline\text{Diverted } & \text {Desviado }\\
\hline\text{CarrierDelay} & \text { Retraso a causa de la aerolínea }\\
\hline\text{WeatherDelay} & \text { Retraso meteorológico }\\
\hline\text{NASDelay} & \text { Retraso a causa del Sistema Aéreo Nacional}\\
\hline\text{SecurityDelay} & \text { Retraso de seguridad}\\
\hline\text{LateAircraftDelay} & \text { Retraso del avión}\\
\hline
\end{array}
\end{aligned}
$$

$$
\begin{aligned}
& \text {Otros Datos}\\
&\begin{array}{|l|r|}
 \hline\text{ActualElapsedTime } & \text { Tiempo real de vuelo }\\
 \hline\text{CRSElapsedTime } & \text { Tiempo de vuelo programado }\\
 \hline\text{AirTime } & \text { Tiempo de vuelo }\\
 \hline\text{Origin } & \text { Origen }\\
\hline\text{Dest } & \text { Destino }\\
\hline\text{Distance } & \text { Distancia }\\
\hline\text{Cancelled} & \text { Cancelado}\\
\hline\text{CancellationCode} & \text {Código de cancelación }\\
\hline
\end{array}
\end{aligned}
$$

* Con el método **.info()** de la libreria pandas, mostraremos un resumen de nuestro dataframe, incluyendo información sobre el tipo de los índices de filas y columnas, los valores no nulos y la memoria usada.

In [48]:
df_flights.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 200000 entries, 365983 to 1537027
Data columns (total 30 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         200000 non-null  int64  
 1   Year               200000 non-null  int64  
 2   Month              200000 non-null  int64  
 3   DayofMonth         200000 non-null  int64  
 4   DayOfWeek          200000 non-null  int64  
 5   DepTime            200000 non-null  float64
 6   CRSDepTime         200000 non-null  int64  
 7   ArrTime            199258 non-null  float64
 8   CRSArrTime         200000 non-null  int64  
 9   UniqueCarrier      200000 non-null  object 
 10  FlightNum          200000 non-null  int64  
 11  TailNum            200000 non-null  object 
 12  ActualElapsedTime  199133 non-null  float64
 13  CRSElapsedTime     199981 non-null  float64
 14  AirTime            199133 non-null  float64
 15  ArrDelay           199133 non-null  float64
 

In [49]:
print('Cancelados',df_flights['Cancelled'].unique())
print('Desviados',df_flights['Diverted'].unique())

Cancelados [0 1]
Desviados [0 1]


* La columna **Cancelled** consta de **0** si el vuelo no ha sido cancelado y **1** caso contrario. Con está variable se puede investigar por ejemplo, el número de vuelos cancelados en el año 2008, vuelos cancelados por compañia....Un razonamiento similar podemos realizar con la columna **Diverted** (vuelos desviados).

* Podeemos prescindir de la columna **Unnamed: 0** pues no aporta nigún tipo de información, si quremos identificar un vuelo lo podemos hacer con la columna **FlightNum** (número de vuelo) la cual tiene un ID único.

## Ejercicio 2
Haga un informe completo del dataset:

1. Resuma estadísticamente el dataset y las columnas de interés. Haga un análisis estadístico del que considere relevante.

In [51]:
df_flights.describe()

Unnamed: 0      Year          Month     DayofMonth      DayOfWeek  \
count  2.000000e+05  200000.0  200000.000000  200000.000000  200000.000000   
mean   3.341257e+06    2008.0       6.110805      15.721995       3.987265   
std    2.067464e+06       0.0       3.485589       8.783651       1.995250   
min    4.200000e+01    2008.0       1.000000       1.000000       1.000000   
25%    1.513686e+06    2008.0       3.000000       8.000000       2.000000   
50%    3.238088e+06    2008.0       6.000000      16.000000       4.000000   
75%    4.975525e+06    2008.0       9.000000      23.000000       6.000000   
max    7.009709e+06    2008.0      12.000000      31.000000       7.000000   

             DepTime     CRSDepTime        ArrTime     CRSArrTime  \
count  200000.000000  200000.000000  199258.000000  200000.000000   
mean     1517.744500    1466.779075    1608.780059    1633.123050   
std       450.598133     425.127599     548.310956     464.971772   
min         1.000000       1.000000       1.000000       0.000000   
25%      1202.000000    1134.000000    1315.000000    1325.000000   
50%      1545.000000    1510.000000    1715.000000    1705.000000   
75%      1900.000000    1815.000000    2029.000000    2014.000000   
max      2400.000000    2359.000000    2400.000000    2400.000000   

           FlightNum  ...       Distance         TaxiIn        TaxiOut  \
count  200000.000000  ...  200000.000000  199258.000000  199961.000000   
mean     2192.579755  ...     765.555975       6.807847      18.238667   
std      1947.094689  ...     574.141263       5.248097      14.507119   
min         1.000000  ...      31.000000       0.000000       0.000000   
25%       619.000000  ...     338.000000       4.000000      10.000000   
50%      1550.000000  ...     606.000000       6.000000      14.000000   
75%      3435.000000  ...     998.000000       8.000000      21.000000   
max      7829.000000  ...    4962.000000     197.000000     321.000000   

           Cancelled       Diverted   CarrierDelay   WeatherDelay  \
count  200000.000000  200000.000000  128947.000000  128947.000000   
mean        0.000290       0.004045      19.287250       3.695131   
std         0.017027       0.063472      44.180207      21.281709   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       2.000000       0.000000   
75%         0.000000       0.000000      21.000000       0.000000   
max         1.000000       1.000000    2436.000000    1352.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  128947.000000  128947.000000      128947.000000  
mean       14.862618       0.085074          25.366484  
std        33.532364       1.851188          42.181637  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         1.000000       0.000000           8.000000  
75%        14.000000       0.000000          33.000000  
max      1289.000000     254.000000        1143.000000  

[8 rows x 25 columns]

* Las variables relevates que podemos considerar son aquellas que están cuatificadas. Por ejemplo, la hora de salida y llegada está representada en formato NASA, los retrasos en minutos, la distancia entre ciudades en millas. Por lo tanto, si quisieramos estudiar los retrasos que ocurren en los aeropuertos y aerolíneas, las columas que contienen estos valores serán muy importantes para desarrollar modelos de predicción.  

2. Encuentre cuántos datos faltantes hay por columna.

In [52]:
print(df_flights.isnull().values.any() )
df_flights.isnull().sum()

True


Unnamed: 0               0
Year                     0
Month                    0
DayofMonth               0
DayOfWeek                0
DepTime                  0
CRSDepTime               0
ArrTime                742
CRSArrTime               0
UniqueCarrier            0
FlightNum                0
TailNum                  0
ActualElapsedTime      867
CRSElapsedTime          19
AirTime                867
ArrDelay               867
DepDelay                 0
Origin                   0
Dest                     0
Distance                 0
TaxiIn                 742
TaxiOut                 39
Cancelled                0
CancellationCode         0
Diverted                 0
CarrierDelay         71053
WeatherDelay         71053
NASDelay             71053
SecurityDelay        71053
LateAircraftDelay    71053
dtype: int64

3. Cree columnas nuevas (velocidad mediana del vuelo, si ha llegado tarde o no...).

* Crear la columna velocidad media de vuelo.

In [53]:
df_flights['Average_velocity']=60*(df_flights['Distance'])/df_flights['ActualElapsedTime']
df_flights.head(5)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
365983      1148586  2008      2          26          2   1321.0        1300   
167859       544978  2008      1          28          1   1909.0        1858   
1203373     4020244  2008      7          31          4   1513.0        1445   
1777761     6566770  2008     12          18          4   1832.0        1822   
432034      1336044  2008      3          15          6   2005.0        1520   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiOut Cancelled  \
365983    1739.0        1718            CO  ...     18.0         0   
167859    2248.0        2238            AS  ...     20.0         0   
1203373   1639.0        1530            MQ  ...     35.0         0   
1777761   2036.0        2020            XE  ...     11.0         0   
432034    2143.0        1646            OH  ...     16.0         0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
365983                  N         0          21.0           0.0       0.0   
167859                  N         0           NaN           NaN       NaN   
1203373                 N         0           0.0           0.0      49.0   
1777761                 N         0          10.0           0.0       6.0   
432034                  N         0         285.0           0.0      12.0   

        SecurityDelay LateAircraftDelay  Average_velocity  
365983            0.0               0.0        401.212121  
167859            NaN               NaN        417.735849  
1203373           0.0              20.0        195.205479  
1777761           0.0               0.0        302.934783  
432034            0.0               0.0        184.285714  

[5 rows x 31 columns]

* Crear la columna si llega tarde o no. Asignaremos el valor **1** si llega tarde y **0** caso contrario.

In [54]:
def ArrLate(x):
    if (x>0):
        return 1
    else:
        return 0

In [55]:
df_flights['ArrLate']=df['ArrDelay'].map(ArrLate);
print('Valores únicos ',df_flights['ArrLate'].unique())
df_flights.head(5)

Valores únicos  [1 0]


Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
365983      1148586  2008      2          26          2   1321.0        1300   
167859       544978  2008      1          28          1   1909.0        1858   
1203373     4020244  2008      7          31          4   1513.0        1445   
1777761     6566770  2008     12          18          4   1832.0        1822   
432034      1336044  2008      3          15          6   2005.0        1520   

         ArrTime  CRSArrTime UniqueCarrier  ...  Cancelled CancellationCode  \
365983    1739.0        1718            CO  ...          0                N   
167859    2248.0        2238            AS  ...          0                N   
1203373   1639.0        1530            MQ  ...          0                N   
1777761   2036.0        2020            XE  ...          0                N   
432034    2143.0        1646            OH  ...          0                N   

         Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
365983          0          21.0           0.0       0.0            0.0   
167859          0           NaN           NaN       NaN            NaN   
1203373         0           0.0           0.0      49.0            0.0   
1777761         0          10.0           0.0       6.0            0.0   
432034          0         285.0           0.0      12.0            0.0   

        LateAircraftDelay Average_velocity  ArrLate  
365983                0.0       401.212121        1  
167859                NaN       417.735849        1  
1203373              20.0       195.205479        1  
1777761               0.0       302.934783        1  
432034                0.0       184.285714        1  

[5 rows x 32 columns]

4. Haga una tabla de las aerolíneas con más atrasos acumulados.

In [56]:
df_Carrier_DepDelay=df[['UniqueCarrier', 'DepDelay']]
df_Carrier_DepDelay.head(5)

UniqueCarrier  DepDelay
0            WN       8.0
1            WN      19.0
2            WN       8.0
3            WN      34.0
4            WN      25.0

In [57]:
flights_depDelay=df_Carrier_DepDelay.groupby('UniqueCarrier').sum()
flights_depDelay.sort_values(by='DepDelay',ascending=False)

DepDelay
UniqueCarrier            
WN             13109808.0
AA              8918260.0
UA              7069279.0
MQ              6205622.0
OO              5931799.0
XE              5193355.0
DL              4462057.0
CO              4322280.0
EV              3955352.0
US              3819499.0
YV              3717346.0
NW              3267642.0
B6              3044016.0
FL              3032082.0
OH              2576892.0
9E              2457810.0
AS              1504939.0
F9               783086.0
HA               248102.0
AQ                20010.0

* La aerolínea con más retrasos acumulados es [WN Southwest Airlines](https://airlinecodes.info/WN).  

5. Cuáles son los vuelos más largos? Y los más atrasados? Buscque las rutas más largas y las que acumulan más retrasos.

####  Vuelos más largos

In [58]:
longest_flights=df_flights[['Origin','Dest','Distance']]
longest_flights=longest_flights.sort_values(by=['Distance'],ascending=False)
longest_flights.drop_duplicates(subset='Distance').head(10)

Origin Dest  Distance
556361     HNL  EWR      4962
1679609    ATL  HNL      4502
1809926    ORD  HNL      4243
1323569    KOA  ORD      4213
447767     ORD  OGG      4184
1884209    MSP  HNL      3972
873289     IAH  HNL      3904
1385557    DFW  HNL      3784
1709863    OGG  DFW      3711
1428668    ANC  ATL      3417

* El vuelos más largo es de Newark (EWR) to Honolulu (HNL) con una distancia de 4962 miles.

#### Vuelos que acumulan más retrasos

In [59]:
Most_delayed_flights=df_flights[['Origin','Dest','DepDelay']]
Most_delayed_flights=Most_delayed_flights.sort_values(by=['DepDelay'],ascending=False)
Most_delayed_flights.drop_duplicates(subset='DepDelay').head(10)


Origin Dest  DepDelay
686014     CLT  MSP    2467.0
545038     LGA  MIA    1521.0
1875741    DSM  ORD    1428.0
1498689    HNL  MSP    1395.0
1599337    SMF  MSP    1351.0
682367     PDX  MSP    1260.0
1909125    DFW  GUC    1178.0
1909638    GUC  DFW    1162.0
1879159    CID  LGA    1134.0
234975     GRR  MCO    1123.0

* El vuelos que acumula más minutos de retraso es:

In [60]:
Most_delayed_flights[Most_delayed_flights.DepDelay==Most_delayed_flights.DepDelay.max()]

Origin Dest  DepDelay
686014    CLT  MSP    2467.0

6. Aporte aquello que consideres relevante.

* Reemplazar los valores faltantes con un cero.

In [61]:
df_flights.fillna(0,inplace=True)
df_flights.isnull().sum()

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
Average_velocity     0
ArrLate              0
dtype: int64

* Unificamos las tres columnas **Year**, **Month** y **DayofMonth** en una sola columnna llamada **Date** de la siguiente manera:

In [62]:
df_flights['Date']=df_flights[['Year','Month','DayofMonth']].apply(lambda x: '{}-{}-{}'.format(x[0], x[1], x[2]), axis=1)
df_flights['Date'] = pd.to_datetime(df_flights['Date'])
df_flights[['Date']].head(5)

Date
365983  2008-02-26
167859  2008-01-28
1203373 2008-07-31
1777761 2008-12-18
432034  2008-03-15

* Eliminamos las columnas **Year**, **Month** y **DayofMonth**. También podemos eliminar la columna **Unnamed: 0**.

In [63]:
df_flights.drop(['Year','Month','DayofMonth', 'Unnamed: 0'],axis=1, inplace=True)
df_flights.head()

DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
365983           2   1321.0        1300   1739.0        1718            CO   
167859           1   1909.0        1858   2248.0        2238            AS   
1203373          4   1513.0        1445   1639.0        1530            MQ   
1777761          4   1832.0        1822   2036.0        2020            XE   
432034           6   2005.0        1520   2143.0        1646            OH   

         FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  ...  \
365983         276  N77296              198.0           198.0  ...   
167859         720  N548AS              159.0           160.0  ...   
1203373       3909  N650AE              146.0           105.0  ...   
1777761       2875  N13975              184.0           178.0  ...   
432034        5478  N594SW               98.0            86.0  ...   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
365983                  N         0          21.0          0.0      0.0   
167859                  N         0           0.0          0.0      0.0   
1203373                 N         0           0.0          0.0     49.0   
1777761                 N         0          10.0          0.0      6.0   
432034                  N         0         285.0          0.0     12.0   

         SecurityDelay  LateAircraftDelay  Average_velocity  ArrLate  \
365983             0.0                0.0        401.212121        1   
167859             0.0                0.0        417.735849        1   
1203373            0.0               20.0        195.205479        1   
1777761            0.0                0.0        302.934783        1   
432034             0.0                0.0        184.285714        1   

              Date  
365983  2008-02-26  
167859  2008-01-28  
1203373 2008-07-31  
1777761 2008-12-18  
432034  2008-03-15  

[5 rows x 29 columns]

* Reiniciar los índices

In [64]:
df_flights.reset_index(drop=True,inplace=True)
df_flights.head()

DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
0          2   1321.0        1300   1739.0        1718            CO   
1          1   1909.0        1858   2248.0        2238            AS   
2          4   1513.0        1445   1639.0        1530            MQ   
3          4   1832.0        1822   2036.0        2020            XE   
4          6   2005.0        1520   2143.0        1646            OH   

   FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  ...  \
0        276  N77296              198.0           198.0  ...   
1        720  N548AS              159.0           160.0  ...   
2       3909  N650AE              146.0           105.0  ...   
3       2875  N13975              184.0           178.0  ...   
4       5478  N594SW               98.0            86.0  ...   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 N         0          21.0          0.0      0.0   
1                 N         0           0.0          0.0      0.0   
2                 N         0           0.0          0.0     49.0   
3                 N         0          10.0          0.0      6.0   
4                 N         0         285.0          0.0     12.0   

   SecurityDelay  LateAircraftDelay  Average_velocity  ArrLate       Date  
0            0.0                0.0        401.212121        1 2008-02-26  
1            0.0                0.0        417.735849        1 2008-01-28  
2            0.0               20.0        195.205479        1 2008-07-31  
3            0.0                0.0        302.934783        1 2008-12-18  
4            0.0                0.0        184.285714        1 2008-03-15  

[5 rows x 29 columns]

* Transformar valores de algunas columnas de formato **float** a **int**

In [65]:
#df_flights=df_flights[[x for x in df_flights.columns if df_flights[x].dtypes==float]].astype('int',copy=False)
float_to_int={'DepTime':int,'ArrTime':int,'CRSDepTime':int,'ActualElapsedTime':int,
              'CRSElapsedTime':int,'DepDelay':int,'TaxiIn':int,'TaxiOut':int,
              'CarrierDelay':int,'WeatherDelay':int,'NASDelay':int,'SecurityDelay':int,
              'LateAircraftDelay':int,'ArrDelay':int,'AirTime':int
             }
df_flights = df_flights.astype(float_to_int)
df_flights
#new_dict = {k:'int' for k in df_flights.columns}
#print(new_dict)

DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
0               2     1321        1300     1739        1718            CO   
1               1     1909        1858     2248        2238            AS   
2               4     1513        1445     1639        1530            MQ   
3               4     1832        1822     2036        2020            XE   
4               6     2005        1520     2143        1646            OH   
...           ...      ...         ...      ...         ...           ...   
199995          7     2052        1858     2304        2119            OO   
199996          4     1456        1405     2031        1930            WN   
199997          4     1153        1120     1458        1420            AA   
199998          1     1233        1120     1500        1335            MQ   
199999          2     1300        1250     1630        1635            WN   

        FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  ...  \
0             276  N77296                198             198  ...   
1             720  N548AS                159             160  ...   
2            3909  N650AE                146             105  ...   
3            2875  N13975                184             178  ...   
4            5478  N594SW                 98              86  ...   
...           ...     ...                ...             ...  ...   
199995       5832  N936SW                 72              81  ...   
199996       3140  N325SW                215             205  ...   
199997       1160  N5ELAA                185             180  ...   
199998       4270  N856AE                147             135  ...   
199999         84  N678AA                150             165  ...   

        CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                      N         0            21            0        0   
1                      N         0             0            0        0   
2                      N         0             0            0       49   
3                      N         0            10            0        6   
4                      N         0           285            0       12   
...                  ...       ...           ...          ...      ...   
199995                 N         0             0            0        0   
199996                 N         0            51            0       10   
199997                 N         0             0            0       38   
199998                 N         0            73            0       12   
199999                 N         0             0            0        0   

        SecurityDelay  LateAircraftDelay  Average_velocity  ArrLate       Date  
0                   0                  0        401.212121        1 2008-02-26  
1                   0                  0        417.735849        1 2008-01-28  
2                   0                 20        195.205479        1 2008-07-31  
3                   0                  0        302.934783        1 2008-12-18  
4                   0                  0        184.285714        1 2008-03-15  
...               ...                ...               ...      ...        ...  
199995              0                105        238.333333        1 2008-05-11  
199996              0                  0        417.767442        1 2008-07-24  
199997              0                  0        355.783784        1 2008-06-05  
199998              0                  0        330.612245        1 2008-07-07  
199999              0                  0        442.000000        0 2008-10-21  

[200000 rows x 29 columns]

In [66]:
df_flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 29 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   DayOfWeek          200000 non-null  int64         
 1   DepTime            200000 non-null  int64         
 2   CRSDepTime         200000 non-null  int64         
 3   ArrTime            200000 non-null  int64         
 4   CRSArrTime         200000 non-null  int64         
 5   UniqueCarrier      200000 non-null  object        
 6   FlightNum          200000 non-null  int64         
 7   TailNum            200000 non-null  object        
 8   ActualElapsedTime  200000 non-null  int64         
 9   CRSElapsedTime     200000 non-null  int64         
 10  AirTime            200000 non-null  int64         
 11  ArrDelay           200000 non-null  int64         
 12  DepDelay           200000 non-null  int64         
 13  Origin             200000 non-null  object  

### Ejercicio 3
Exporte el dataset limpio y con las nuevas columnas a Excel.

In [71]:
#dfxlsx=pd.read_excel('Flights.xlsx')
#dfxlsx.head()

DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
0          2     1321        1300     1739        1718            CO   
1          1     1909        1858     2248        2238            AS   
2          4     1513        1445     1639        1530            MQ   
3          4     1832        1822     2036        2020            XE   
4          6     2005        1520     2143        1646            OH   

   FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  ...  \
0        276  N77296                198             198  ...   
1        720  N548AS                159             160  ...   
2       3909  N650AE                146             105  ...   
3       2875  N13975                184             178  ...   
4       5478  N594SW                 98              86  ...   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 N         0            21            0        0   
1                 N         0             0            0        0   
2                 N         0             0            0       49   
3                 N         0            10            0        6   
4                 N         0           285            0       12   

   SecurityDelay  LateAircraftDelay  Average_velocity  ArrLate       Date  
0              0                  0        401.212121        1 2008-02-26  
1              0                  0        417.735849        1 2008-01-28  
2              0                 20        195.205479        1 2008-07-31  
3              0                  0        302.934783        1 2008-12-18  
4              0                  0        184.285714        1 2008-03-15  

[5 rows x 29 columns]